File: .github/cross-shim-tests/worker.ts
// Cross-shim wire-format test fixture: Node worker.
//
// Mirrors `worker.py`: drains COUNT distinct jobs from QUEUE, asserts
// each payload is `{ i: int in [0, COUNT), tag: EXPECT_TAG }`, and
// exits 0 on full coverage within TIMEOUT_SECS, else 1.
//
// EXPECT_JOB_NAME — optional. When non-empty, the handler asserts
// `job.name === EXPECT_JOB_NAME` so a regression that drops `name`
// on either shim's wire path is caught here.
//
// STORE_RESULT  — optional. When '1', the worker enables `storeResults`
//                 so the engine persists each handler's return value at
//                 `{chasqui:<QUEUE>}:result:<jobId>` for the verifier to
//                 read back. Default off.
// RESULT_VALUE  — optional, JSON-encoded. When set, the handler returns
//                 `JSON.parse(RESULT_VALUE)` instead of `undefined`.
// EMIT_PROGRESS — optional. When '1' the handler calls
//                 `job.updateProgress(75)` and appends two log lines
//                 ('step 1', 'step 2') before returning so a
//                 `verify_progress` step in the opposite shim can read
//                 them back via `Queue.getJob` / `Queue.getJobLogs`.
//                 Pair with `STORE_RESULT=1` to keep completed jobs
//                 discoverable by the introspector.
//
// LIMITER_MAX          — optional. When set (with LIMITER_DURATION_MS), the
//                        worker enables a global per-queue rate limiter of
//                        `max` jobs per `duration` ms window (shared Redis
//                        bucket). Used by the rate-limit cross-shim phase to
//                        prove the Node + Python FFI paths share ONE bucket.
// LIMITER_DURATION_MS  — optional. Window length in ms; required when
//                        LIMITER_MAX is set.

import { Worker } from '../../chasquimq-node/dist/index.js'

async function main(): Promise<number> {
  const queueName = requireEnv('QUEUE')
  const count = Number(requireEnv('COUNT'))
  const expectTag = process.env.EXPECT_TAG ?? 'node'
  const expectJobName = process.env.EXPECT_JOB_NAME ?? ''
  const timeoutSecs = Number(process.env.TIMEOUT_SECS ?? '30')
  const redisUrl = process.env.REDIS_URL ?? 'redis://127.0.0.1:6379'
  const storeResults = process.env.STORE_RESULT === '1'
  const resultValueRaw = process.env.RESULT_VALUE ?? ''
  const resultValue: unknown = resultValueRaw
    ? JSON.parse(resultValueRaw)
    : undefined
  const emitProgress = process.env.EMIT_PROGRESS === '1'
  const limiterMaxRaw = process.env.LIMITER_MAX ?? ''
  const limiterDurationRaw = process.env.LIMITER_DURATION_MS ?? ''
  const limiter = limiterMaxRaw
    ? { max: Number(limiterMaxRaw), duration: Number(limiterDurationRaw) }
    : undefined

  const seen = new Set<number>()
  const errors: string[] = []
  let resolveDone: () => void
  const done = new Promise<void>((r) => {
    resolveDone = r
  })

  const worker = new Worker<{ i: number; tag: string }, unknown>(
    queueName,
    async (job) => {
      const data = job.data as unknown
      if (typeof data !== 'object' || data === null) {
        errors.push(`payload not an object: ${JSON.stringify(data)}`)
        resolveDone()
        return resultValue
      }
      const { i, tag } = data as { i?: unknown; tag?: unknown }
      if (typeof i !== 'number' || !Number.isInteger(i) || i < 0 || i >= count) {
        errors.push(`i out of range: ${JSON.stringify(i)}`)
        resolveDone()
        return resultValue
      }
      if (tag !== expectTag) {
        errors.push(`tag mismatch: got ${JSON.stringify(tag)}, want '${expectTag}'`)
        resolveDone()
        return resultValue
      }
      if (expectJobName && job.name !== expectJobName) {
        errors.push(`name mismatch: got '${job.name}', want '${expectJobName}'`)
        resolveDone()
        return resultValue
      }
      if (emitProgress) {
        try {
          await job.updateProgress(75)
          await job.log('step 1')
          await job.log('step 2')
        } catch (err) {
          errors.push(`emitProgress failed: ${(err as Error).message}`)
          resolveDone()
          return resultValue
        }
      }
      seen.add(i)
      if (seen.size >= count) {
        resolveDone()
      }
      return resultValue
    },
    {
      connection: parseConn(redisUrl),
      concurrency: 8,
      autorun: false,
      maxStalledCount: 1,
      drainDelay: 200,
      runScheduler: false,
      storeResults,
      limiter,
    },
  )

  void worker.run()

  let timedOut = false
  const timer = setTimeout(() => {
    timedOut = true
    resolveDone!()
  }, timeoutSecs * 1000)

  try {
    await done
  } finally {
    clearTimeout(timer)
    await worker.close().catch(() => {})
  }

  if (timedOut && seen.size < count) {
    console.error(
      `[node-worker] TIMEOUT after ${timeoutSecs}s — saw ${seen.size}/${count}`,
    )
    return 1
  }
  if (errors.length > 0) {
    for (const e of errors) console.error(`[node-worker] ERROR: ${e}`)
    return 1
  }
  if (seen.size !== count) {
    const missing = []
    for (let i = 0; i < count && missing.length < 10; i++) {
      if (!seen.has(i)) missing.push(i)
    }
    console.error(
      `[node-worker] coverage gap: saw ${seen.size}/${count} (missing: ${missing.join(',')}...)`,
    )
    return 1
  }

  console.log(
    `[node-worker] OK — drained ${count} distinct jobs with tag='${expectTag}' name='${expectJobName}'`,
  )
  return 0
}

function requireEnv(name: string): string {
  const v = process.env[name]
  if (v == null || v === '') {
    throw new Error(`missing required env var ${name}`)
  }
  return v
}

function parseConn(url: string) {
  const u = new URL(url)
  return {
    host: u.hostname || '127.0.0.1',
    port: u.port ? Number(u.port) : 6379,
    password: u.password || undefined,
    username: u.username || undefined,
    db: u.pathname && u.pathname !== '/' ? Number(u.pathname.slice(1)) : undefined,
  }
}

main().then(
  (code) => process.exit(code),
  (err) => {
    console.error(err)
    process.exit(1)
  },
)

File: .github/cross-shim-tests/worker.py
"""Cross-shim wire-format test fixture: Python worker.

Consumes COUNT jobs from QUEUE. Each payload must be a dict with
integer ``i`` in [0, COUNT) and string ``tag`` matching EXPECT_TAG.
Exits 0 on full distinct-id coverage within TIMEOUT_SECS, else 1.

The worker leaves the engine default ``delayed_enabled=True`` regardless
of the producer's MODE — a worker should drain whatever's available, so
gating the embedded promoter on a producer-side env var creates a fragile
two-process coupling (forget MODE on the worker side and the ZSET sits
forever). This mirrors the Node fixture's approach.

Env vars:
  QUEUE, COUNT — required.
  EXPECT_JOB_NAME  — optional. When non-empty, the handler asserts
                      ``job.name == EXPECT_JOB_NAME`` for every job, so
                      a regression that drops `name` on either shim's
                      wire path is caught here.
  STORE_RESULT  — optional. When ``"1"`` the worker enables
                   ``store_results=True`` so the engine persists the
                   handler's return value at
                   ``{chasqui:<QUEUE>}:result:<jobId>`` for the verifier
                   to read back. Default off.
  RESULT_VALUE  — optional, JSON-encoded. When set, the handler returns
                   ``json.loads(RESULT_VALUE)`` instead of ``None``. The
                   verifier asserts this value round-trips through the
                   shim's msgpack wire format.
  EXPECT_TAG, TIMEOUT_SECS, REDIS_URL — optional.
"""

from __future__ import annotations

import asyncio
import json
import os
import sys

from chasquimq import Job, Worker


async def main() -> int:
    queue_name = os.environ["QUEUE"]
    count = int(os.environ["COUNT"])
    expect_tag = os.environ.get("EXPECT_TAG", "py")
    expect_job_name = os.environ.get("EXPECT_JOB_NAME", "")
    timeout_secs = float(os.environ.get("TIMEOUT_SECS", "30"))
    redis_url = os.environ.get("REDIS_URL", "redis://127.0.0.1:6379")
    store_results = os.environ.get("STORE_RESULT", "") == "1"
    result_value_raw = os.environ.get("RESULT_VALUE", "")
    result_value = json.loads(result_value_raw) if result_value_raw else None

    seen: set[int] = set()
    done = asyncio.Event()
    errors: list[str] = []

    async def handler(job: Job):
        data = job.data
        if not isinstance(data, dict):
            errors.append(f"payload not a dict: {data!r}")
            done.set()
            return None
        i = data.get("i")
        tag = data.get("tag")
        if not isinstance(i, int) or i < 0 or i >= count:
            errors.append(f"i out of range: {i!r}")
            done.set()
            return None
        if tag != expect_tag:
            errors.append(f"tag mismatch: got {tag!r}, want {expect_tag!r}")
            done.set()
            return None
        if expect_job_name and job.name != expect_job_name:
            errors.append(
                f"name mismatch: got {job.name!r}, want {expect_job_name!r}"
            )
            done.set()
            return None
        seen.add(i)
        if len(seen) >= count:
            done.set()
        return result_value

    worker = Worker(
        queue_name,
        handler,
        redis_url=redis_url,
        concurrency=8,
        max_attempts=1,
        read_block_ms=200,
        run_scheduler=False,
        store_results=store_results,
    )

    run_task = asyncio.create_task(worker.run())
    try:
        try:
            await asyncio.wait_for(done.wait(), timeout=timeout_secs)
        except asyncio.TimeoutError:
            print(
                f"[py-worker] TIMEOUT after {timeout_secs}s — saw {len(seen)}/{count}",
                file=sys.stderr,
            )
            return 1
    finally:
        await worker.close()
        try:
            await asyncio.wait_for(run_task, timeout=5.0)
        except asyncio.TimeoutError:
            pass
        await asyncio.sleep(0)

    if errors:
        for e in errors:
            print(f"[py-worker] ERROR: {e}", file=sys.stderr)
        return 1
    if len(seen) != count:
        print(
            f"[py-worker] coverage gap: saw {len(seen)}/{count} (missing: "
            f"{sorted(set(range(count)) - seen)[:10]}...)",
            file=sys.stderr,
        )
        return 1

    print(
        f"[py-worker] OK — drained {count} distinct jobs with "
        f"tag={expect_tag!r} name={expect_job_name!r}"
    )
    return 0


if __name__ == "__main__":
    sys.exit(asyncio.run(main()))
